consensus/service/src/client_api_service.rs

```diff
@@ -158,3 +157,87 @@ impl<E: ConsensusEnclaveProxy, L: Ledger + Clone> ConsensusClientApi for ClientA
         });
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{client_api_service::ClientApiService, tx_manager::MockTxManager};
+    use grpcio::{ChannelBuilder, Environment, ServerBuilder};
+    use mc_attest_api::attest::Message;
+    use mc_common::{
+        logger::{test_with_logger, Logger},
+        NodeID, ResponderId,
+    };
+    use mc_consensus_api::{
+        consensus_client_grpc, consensus_client_grpc::ConsensusClientApiClient,
+        consensus_common::ProposeTxResult,
+    };
+    use mc_consensus_enclave::TxContext;
+    use mc_consensus_enclave_mock::MockConsensusEnclave;
+    use mc_transaction_core::tx::TxHash;
+    use mc_transaction_core_test_utils::{create_ledger, initialize_ledger, AccountKey};
+    use rand::{rngs::StdRng, SeedableRng};
+    use std::sync::Arc;
+
+    #[test_with_logger]
+    fn test_client_tx_propose_ok(logger: Logger) {
+        let mut consensus_enclave = MockConsensusEnclave::new();
+        consensus_enclave
+            .expect_client_tx_propose()
+            .times(1)
+            .return_const(Ok(TxContext::default()));
+
+        // Arc<dyn Fn(TxHash, Option<&NodeID>, Option<&ResponderId>) + Sync + Send>
+        let scp_client_value_sender = Arc::new(
+            |_tx_hash: TxHash, _node_id: Option<&NodeID>, _responder_id: Option<&ResponderId>| {
+                // TODO: store inputs for inspection.
+            },
+        );
+
+        // Local ledger
+        let mut ledger = create_ledger();
+        let mut rng: StdRng = SeedableRng::from_seed([62u8; 32]);
+        let sender = AccountKey::random(&mut rng);
+        let num_blocks = 5;
+        initialize_ledger(&mut ledger, num_blocks, &sender, &mut rng);
+
+        let mut tx_manager = MockTxManager::new();
+        tx_manager
+            .expect_insert()
+            .times(1)
+            .return_const(Ok(TxHash::default()));
+
+        let is_serving_fn = Arc::new(|| -> bool { true });
+
+        let client_api_service = ClientApiService::new(
+            Arc::new(consensus_enclave),
+            scp_client_value_sender,
+            ledger,
+            Arc::new(tx_manager),
+            is_serving_fn,
+            logger,
+        );
+
+        let service = consensus_client_grpc::create_consensus_client_api(client_api_service);
+
+        let env = Arc::new(Environment::new(1));
+        let mut server = ServerBuilder::new(env.clone())
+            .register_service(service)
+            .bind("127.0.0.1", 0)
```

Comment:
Unit tests like this might need to run sequentially.

Reply:
Why is that? Shouldn't it pick an available port every time?

Reply:
https://github.com/mobilecoinofficial/mobilecoin/blob/master/mobilecoind/src/test_utils.rs#L218 ?

Reply:
That also works